target/linux/ramips/dts/mt7621_xiaomi_mi-router-4.dts

```diff
@@ -0,0 +1,200 @@
+			gpios = <&gpio 10 GPIO_ACTIVE_LOW>;
+		};
+
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		minet {
+			label = "minet";
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+	};
+
+};
+
+&nand {
+	status = "okay";
+
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "Bootloader";
+			reg = <0x0 0x80000>;
+			read-only;
+		};
+
+		partition@80000 {
+			label = "Config";
+			reg = <0x80000 0x40000>;
+		};
+
+		partition@c0000 {
+			label = "Bdata";
+			reg = <0xc0000 0x40000>;
+			read-only;
+		};
+
+		factory: partition@100000 {
+			label = "factory";
+			reg = <0x100000 0x40000>;
+			read-only;
+		};
+
+		partition@140000 {
+			label = "crash";
+			reg = <0x140000 0x40000>;
+		};
+
+		partition@180000 {
+			label = "crash_syslog";
+			reg = <0x180000 0x40000>;
+		};
+
+		partition@1c0000 {
+			label = "reserved0";
+			reg = <0x1c0000 0x40000>;
+			read-only;
+		};
+
+		/* uboot expects to find kernels at 0x200000 & 0x600000
+		 * referred to as system 1 & system 2 respectively.
+		 * a kernel is considered suitable for handing control over
+		 * if its linux magic number exists & uImage CRC are correct.
+		 * If either of those conditions fail, a matching sys'n'_fail flag
+		 * is set in uboot env & a restart performed in the hope that the
+		 * alternate kernel is okay.
+		 * if neither kernel checksums ok and both are marked failed, system 2
+		 * is booted anyway.
+		 *
+		 * Note uboot's tftp flash install writes the transferred
+		 * image to both kernel partitions.
+		 */
+
+		partition@200000 {
+			label = "kernel_stock";
+			reg = <0x200000 0x400000>;
+		};
+
+		partition@600000 {
+			label = "kernel";
+			reg = <0x600000 0x400000>;
+		};
+
+		/* ubi partition is the result of squashing
+		 * next consecutive stock partitions:
+		 * - rootfs0 (rootfs partition for stock kernel0),
+		 * - rootfs1 (rootfs partition for stock failsafe kernel1),
+		 * - overlay (used as ubi overlay in stock fw)
+		 * resulting 117,5MiB space for packages.
+		 */
+
+		partition@a00000 {
+			label = "ubi";
+			reg = <0xa00000 0x7580000>;
+		};
+	};
+};
+
+&pcie {
+	status = "okay";
+};
+
+&pcie0 {
+	wifi@0,0 {
+		compatible = "pci14c3,7603";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
+		ieee80211-freq-limit = <2400000 2500000>;
+	};
+};
+
+&pcie1 {
+	wifi@0,0 {
+		compatible = "pci14c3,7662";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x8000>;
+		ieee80211-freq-limit = <5000000 6000000>;
+	};
+};
+
+&gmac0 {
+	mtd-mac-address = <&factory 0xe006>;
+};
+
+&switch0 {
+	ports {
+		port@2 {
+			status = "okay";
+			label = "lan2";
+		};
+
+		port@1 {
+			status = "okay";
+			label = "lan1";
+		};
+
+		port@4 {
+			status = "okay";
+			label = "wan";
+			mtd-mac-address = <&factory 0xe000>;
```

Comment:
Do LAN/WAN MAC addresses match vendor firmware assignment.

Reply:
Yes, _typically_. But even better would be to check what the vendor is actually doing for this particular device.

Apart from that, this change also affected the label MAC address derived via label-mac-device in DTS. Only one version can be correct here, the old one or the new one, as the label shows a specific address. Which one is it, that of LAN or WAN (and for which configuration)?

Reply:
```
root@OpenWrt:~# ip addr
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN qlen 1000
    link/ether 40:31:3c:14:63:c2 brd ff:ff:ff:ff:ff:ff
5: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether 40:31:3c:14:63:c3 brd ff:ff:ff:ff:ff:ff
6: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether 40:31:3c:14:63:c3 brd ff:ff:ff:ff:ff:ff
7: eth0.2@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether 40:31:3c:14:63:c2 brd ff:ff:ff:ff:ff:ff
8: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether 40:31:3c:14:63:c4 brd ff:ff:ff:ff:ff:ff
9: wlan1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether 40:31:3c:14:63:c5 brd ff:ff:ff:ff:ff:ff

mtd3: 00040000 00020000 "Factory"

root@OpenWrt:~# hexdump -C /dev/mtd3
00000000  03 76 01 02 40 31 3c 14  63 c4 03 76 c3 14 ff ff  |.v..@1<.c..v....|
...
00008000  62 76 01 00 40 31 3c 14  63 c5 12 76 c3 14 00 00  |bv..@1<.c..v....|
...
0000e000  40 31 3c 14 63 c2 40 31  3c 14 63 c3 ff ff ff ff  |@1<.c.@1<.c.....|

root@OpenWrt:~# find /proc/device-tree/ -name "*mac-address"
/proc/device-tree/ethernet@1e100000/mac-address
/proc/device-tree/ethernet@1e100000/mtd-mac-address
root@OpenWrt:~#
root@OpenWrt:~# . /lib/functions/system.sh
root@OpenWrt:~# get_mac_binary "/proc/device-tree/ethernet@1e100000/mac-address" 0
40:31:3c:14:63:c2
root@OpenWrt:~# get_mac_binary "/proc/device-tree/ethernet@1e100000/mtd-mac-address" 0
00:00:00:0d:00:00
```
I'm running @ioiotor build and has used next:
https://github.com/ioiotor/mir4-ss/blob/master/target/linux/ramips/dts/MIR4.dts
```
&ethernet {
	mtd-mac-address = <&factory 0xe000>;
	mediatek,portmap = "llllw";
};
```
Hope I found them correctly. Can you help to update dts accordingly?

Reply:
So, this is just the data from another OpenWrt fork?

Reply:
Yes. Since mir3g is the same as mir4 except for lower RAM and additional gpio button, I suppose it can be copied from there?

Reply:
According to u-boot log 
mir3g's portmap = "lwlll"; 
mir4's portmap = "llllw";
It's different 

Sorry, I'm not at home right now. I can't provide u-boot logs